src/impl-mezzano.lisp

```diff
@@ -0,0 +1,137 @@
+;;;; -*- indent-tabs-mode: nil -*-
+
+#|
+Copyright 2006, 2007 Greg Pfeil
+Copyright 2016 Henry Harrington
+
+Distributed under the MIT license (see LICENSE file)
+|#
+
+(in-package #:bordeaux-threads)
+
+(deftype thread ()
+  'mezzano.supervisor:thread)
+
+;;; Thread Creation
+
+(defun %make-thread (function name)
+  (mezzano.supervisor:make-thread
+   (lambda ()
+     (funcall function))
+   :name name))
+
+(defun current-thread ()
+  (mezzano.supervisor:current-thread))
+
+(defun threadp (object)
+  (mezzano.supervisor:threadp object))
+
+(defun thread-name (thread)
+  (mezzano.supervisor:thread-name thread))
+
+;;; Resource contention: locks and recursive locks
+
+(deftype lock () 'mezzano.supervisor:mutex)
+
+(defun lock-p (object)
+  (mezzano.supervisor:mutex-p object))
+
+(defun make-lock (&optional name)
+  (mezzano.supervisor:make-mutex name))
+
+(defun acquire-lock (lock &optional (wait-p t))
+  (mezzano.supervisor:acquire-mutex lock wait-p))
+
+(defun release-lock (lock)
+  (mezzano.supervisor:release-mutex lock))
+
+(defmacro with-lock-held ((place) &body body)
+  `(mezzano.supervisor:with-mutex (,place) ,@body))
+
+(defstruct (recursive-lock
+             (:constructor make-recursive-lock
+                           (&optional name &aux
+                                      (mutex (mezzano.supervisor:make-mutex name)))))
+  mutex
+  (depth 0))
+
+(defun call-with-recursive-lock-held (lock function)
+  (cond ((mezzano.supervisor:mutex-held-p
+          (recursive-lock-mutex lock))
+         (unwind-protect
+              (progn (incf (recursive-lock-depth lock))
+                     (funcall function))
+           (decf (recursive-lock-depth lock))))
+        (t
+         (mezzano.supervisor:with-mutex ((recursive-lock-mutex lock))
+           (multiple-value-prog1
+               (funcall function)
+             (assert (eql (recursive-lock-depth lock) 0)))))))
+
+(defmacro with-recursive-lock-held ((place) &body body)
+  `(call-with-recursive-lock-held ,place (lambda () ,@body)))
+
+;;; Resource contention: condition variables
+
+(defun make-condition-variable (&key name)
+  (mezzano.supervisor:make-condition-variable name))
+
+(defun condition-wait (condition-variable lock &key timeout)
+  (mezzano.supervisor:condition-wait
+   condition-variable
+   (if (recursive-lock-p lock)
```

Comment:
Conditions should always use regular mutexes.

Reply:
I don't understand.

Reply:
The code shouldn't check for recursive locks because condition variables should only use regular locks.